### DISCUSSION TITLE
[일지] 5월 28일 (금) 오후 회의록

### DISCUSSION THREAD

**Discussion by olenmg** [bcaitech1/p4-opt-5-vibrhanium-#17] [일지]
### 조용래 멘토님 멘토링 내용
OCR, 음성,번역 모델쪽에서 현업 종사하고 계심

- 파이썬은 병렬처리가 힘들어서 모델 서빙에 잘 사용되지 않음.
- 모델 서빙은 Java (Scala) 로 했었음. (모델 개발은 TF로 했었음)
- TF, pytorch 내부는 C++. 모델 서빙을 할 때는 C++과의 인터페이스를 만들어서 TF - C++ - Java 로 서빙함.
- 요즘에는 pytorch나 TF의 checkpoint를 다른 언어 (C, Java 등)에서 다룸. 공식적으로 지원해줌. 
- PYTHON은 엔진 역할로만, ??? 
- Quantization 관련 LIBRARY 가 이미 있어서 이걸 가져다 쓰는 방법 추천
- 일반적으로 많이 사용되는 모델에서 출발
    - mobilenet은 이미 compact해서 prunning은 효과가 미약할 수 있다.
    - quntization을 적용하면 더 경량화 가능할 수 있다.

- AutoML 학습 시간을 줄이는 방법: 적은 데이터 input을 사용한 다음에 최적의 하이퍼 파라미터를 큰 데이터셋에 적용
- 멘토링 시간: 목요일 20시

### 멘토님께 질문 
1. autoML로 신경망을 구성할 때 기본 구조는 어떻게 설계되나요?
    - 모델 레벨까지 automl로 돌리는건 아직 연구 레벨, 보통은 hyperparameter tuning 시에 많이 활용
    - 모델 구조에 대한 automl 보다는 hyperparameter tuning에 automl을 활용하는게 좀 더 안정적인 방법
    - 만약 대회에서 하이퍼파라미터 튜닝이 적당히 이루어졌다고 생각되면, 그때가서 모델 구조에 대한 튜닝을 생각해보는게 좋을 것
    - 그 외에, pruning, quantization은 라이브러리를 통해 활용, 이론을 접하고 라이브러리 활용을 익히는 것도 하나의 학습이 될 것
2. batch size, 모델 구조, loss, lr_scheduler, optimizer등 어떤 걸 우선적으로 찾으시는지 궁금합니다 : ) 
    - 잘 돌아가는 모델을 먼저 찾고, 일반적으로 많이 사용되는 모델: yolo, resent/resnext, imageNet
    - 상황에 따라서 다르지만, LR이 요즘 제일 중요하다고 생각한다. (automl 적용시)
    - batch size는 최대
    - optimizer는 대부분 Adam을 많이 활용. 요즘은 AdamW, AdamP 등 더 발전된 adam optimizer가 많이 나왔으나 성능 차이는 크게 없긴 함
    - 모델 구조는 아이디어에 따라 다르다
    - loss를 바꿔보는 것은 새로운 정보(pseudo label 등) 추가 or 공식 변경 => 베이스라인과 우리가 풀려는 문제가 서로 다른 경우에 적용해볼만 함
    - 논문에서 활용하고 있는 데이터와 우리 대회에서의 데이터가 다른 경우에 이런걸 고려해볼만함. 모델별 논문 읽어보는게 좋을듯. 어떤 데이터를 쓰는지
    - multi-task learning을 응용하여 사용하는것도 성능에 도움이 된다.
3. 멘토님이 경량화 대회 참여하신다면 어떤 전략을 취하실 것 같으신가요?
    - 이전에 답변
4. (후보) 현업에서도 AutoML로 모델을 만들어서 사용하는 경우가 있나요?
    - 현업에서도 활용하는 경우가 있다.
    - 컴퓨팅 파워가 많이 들어 많이 사용하지는 못한다.
    - 대기업에서는 많이 활용(네이버 - NSML(AutoML 기능이 내장됨))
    - 1% ~ 2% 성능향상도 중요할 수 있으므로 이를 위해 활용한다.
    - 스타트업에서는 많이 활용하지는 않음. 
5. (후보) Sample data를 뽑아서 실험에 사용하고 싶은데, sample data를 뽑으시는 방법이 있으신가요? ex. train data와 동일한 Category별 분포를 유지하며, random으로 추출
    - random으로 많이 활용하는데, 웬만하면 실제와 분포가 비슷한(전체 데이터와 분포가 비슷한) 데이터를 뽑아서 해보는 게 좋음
6. 하이퍼 클로바(hyperCLOVA)에 대한 현업에 있는 분의 생각을 듣고 싶습니다.
    - ~~ 
7. CV,NLP 모집은 봤는데 경량화ML은 어떻게 가나요..!?
    - 경량화에 대한 기업의 수요는 많다.
    - 경량화는 모델에 대한 이해가 있는 사람이 경량화를 할 수 있다. 
    - 모델 경량화보다 HW 경량화가 더 쉽다. ? Distillation, Quantization 시에는 성능보장이 없다.
8. AutoML을 통해 hyperparamteer를 **빠르게** 찾는 방법은?
    - input 데이터를 좀 줄인다
        - 적은 데이터를 베이스라인으로 활용하고 대충 search space를 찾아 더 많은 데이터로 확장
9. 기업이 지금까지의 경험을 봤을 때 어떻게 볼지, 우리가 어떤 식으로 어필을 해야하는지?
    - 회사마다 보는 부분이 다르다. 
        - application 능력 & 툴 활용 능력 보는 팀
        - fundamental한 ML 지식, 수학지식을 깊게 보는 팀
        - ★★★ 지원자가 진행한 프로젝트에 대해 묻는 경우 어디까지 고민을 해보았는가?
        - 기초적인 지식: batch normalization, attention / 베이즈정리 등 을 묻는 팀
        - 팀에 따라 강조하는 포인트는 다르다

### 예정
- 다음주 월요일 Special mission 3, 4 진행
    - 재경, 건우 발표 예정(방향 다르게)

- 월요일 피어세션 5시 진행
